### PR TITLE
loopdb: add Secret type for reading password from file

### LINKED ISSUE #1088: Add Secret type for reading database password from file
## Background

This issue tracks a follow-up suggestion from PR #1078: staticaddr: various fixes

**Original Comment**: https://github.com/lightninglabs/loop/pull/1078#discussion_r2894023390
**Author**: @starius

> _Nit for a follow-up PR if you like it._
>
> I wish flags' parsing libraries provided a convenient way to pass password as files. Like:
>
> ```go
> Password  string `long:"password" description:"Database user's password." from:"file"`
> ```
>
> Unfortunately it is not supported, but we can make a custom field type:
>
> ```go
>   type Secret string
>
>   func (s *Secret) UnmarshalFlag(v string) error {
>         // Example convention: @/path/to/file => read from file, else raw value
>         if strings.HasPrefix(v, "@") {
>                 b, err := os.ReadFile(strings.TrimPrefix(v, "@"))
>                 if err != nil {
>                         return err
>                 }
>                 *s = Secret(strings.TrimRight(string(b), "\r\n"))
>                 return nil
>         }
>         *s = Secret(v)
>         return nil
>   }
>
>   type Opts struct {
>         Password   Secret `long:"password" description:"Database user's password or @file to read the password from it."`
>   }
> ```

## Context

PR #1078 added a `//nolint:gosec` comment to suppress a linter warning about the `Password` field in `PostgresConfig`. The comment suggests a better long-term solution: instead of just suppressing the lint warning, introduce a custom `Secret` type that supports reading passwords from files, avoiding plaintext passwords in CLI arguments or config files.

## Proposed Change

Introduce a `Secret` string type that implements `UnmarshalFlag` with a convention where values prefixed with `@` are treated as file paths. This allows users to pass `--password @/path/to/secret` to read the password from a file, while still supporting raw string values for backward compatibility.

This is a security improvement — passing secrets via files avoids leaking them through process listings (`ps aux`) and shell history.

## Implementation Approach

- Add a `Secret` type in an appropriate package (e.g., `loopdb` or a shared `config` package).
- Implement `UnmarshalFlag(string) error` on `*Secret` so `go-flags` calls it automatically.
- If value starts with `@`, read the file at the remaining path and trim trailing newlines.
- Otherwise, use the raw value directly.
- Change `PostgresConfig.Password` from `string` to `Secret` in `loopdb/postgres.go`.
- Update `DSN()` and any other callers to cast `Secret` back to `string` where needed.
- Update the field description to document the `@file` convention.
- The `//nolint:gosec` annotation can likely be removed since the field type is no longer a raw `string` named `Password`.

## Related

- PR #1078

### PR DESCRIPTION
**Pull request by kaldun-tech**
## Summary

- Add `Secret` type that implements `go-flags.Unmarshaler` to support reading sensitive values from files using `@/path/to/file` syntax
- Change `PostgresConfig.Password` field from `string` to `Secret`
- This allows the postgres password to be stored in a file rather than passed directly on the command line, avoiding exposure in process listings and shell history

## Usage

```bash
# Direct password (existing behavior)
loopd --postgres.password=mypassword

# Password from file (new behavior)
echo "mypassword" > /path/to/secret.txt
chmod 600 /path/to/secret.txt
loopd --postgres.password=@/path/to/secret.txt
```

Trailing whitespace (spaces, tabs, newlines) is automatically stripped from file contents.

## Design notes

- **Ecosystem precedent**: This is a new pattern for Lightning Labs projects. lnd uses a single DSN connection string; taproot-assets uses a plain string Password field. This Secret type could serve as a template for similar enhancements in other projects.
- **No `fmt.Stringer`/`fmt.GoStringer` implementation**: The Secret type does not implement these interfaces. This is intentional because `PostgresConfig.DSN(hidePassword bool)` already handles password redaction explicitly. Adding automatic redaction via Stringer could mask debugging issues and create a false sense of security if   the value is printed through other means (e.g., `%v` on a containing struct).
- **No `MarshalFlag`**: The resolved secret should not be written back to config files. Users should store the `@/path` syntax, not the resolved value.
- **Error messages**: include the file path for debugging, ex. `secret file not found: /path/to/file`
- **Whitespace handling**: Trailing whitespace (spaces, tabs, `\r`, `\n`) is stripped from file contents
- **No escape syntax for literal `@` prefix**: Passwords that start with `@` are interpreted as file paths.

## Test plan

- [x] Unit tests for Secret type covering direct values, file references, edge cases
- [x] go-flags integration test confirming parser compatibility
- [x] Manual test: `loopd --postgres.password=@/path/to/secret.txt` parses successfully
- [x] Lint: `make lint` or `golangci-lint run ./loopdb/...`

Fixes #1088

### CODE REVIEW

File: loopdb/postgres.go [hunk 1]
```diff
@@ -33,7 +33,7 @@ type PostgresConfig struct {
 	Host               string `long:"host" description:"Database server hostname."`
 	Port               int    `long:"port" description:"Database server port."`
 	User               string `long:"user" description:"Database user."`
-	Password           string `long:"password" description:"Database user's password."` //nolint:gosec
+	Password           Secret `long:"password" description:"Database user's password. Use @/path/to/file to read from a file."`
```
[the remark this answers is not in the document]
Comment (kaldun-tech) [reply loopdb/postgres.go]: I think this can be ignored: The gosec G101 rule looks for hardcoded credentials in string literals. Since Secret is a type (not a hardcoded value), it shouldn't trigger. 
Running the linter showed no warnings in this file via ```golangci-lint run -v```

File: loopdb/secret.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"fmt"
+	"os"
+	"strings"
+)
+
+// Secret is a string type that can unmarshal values from files when prefixed
+// with '@'. This allows sensitive values like passwords to be stored in files
+// rather than directly in configuration.
+type Secret string
```
[the remark this answers is not in the document]
Comment (kaldun-tech) [reply loopdb/secret.go:12]: This is a deliberate design decision made because:
- DSN() already handles password hiding explicitly via the hidePassword parameter
- Automatic redaction could mask debugging issues
- This keeps the Secret type simple and predictable

File: loopdb/secret.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+					"file (permission denied): %s",
+					filePath)
+			}
+
+			return fmt.Errorf("failed to read secret file %s: %w",
+				filePath, err)
+		}
+		// Trim trailing newlines. Trim both \r and \n to handle
+		// files created on Windows (CRLF) or Unix (LF).
+		*s = Secret(strings.TrimRight(string(content), "\r\n"))
```
[the remark this answers is not in the document]
Comment (kaldun-tech) [reply loopdb/secret.go]: I will make this update. The downside is a password intentionally ending with a space would break (edge case)

**Comment by kaldun-tech** [comment]
@hieblmi @bhandras @alexbosworth @starius I believe this is ready for review.

**Comment by kaldun-tech** [comment]
@hieblmi @bhandras Ready for review

**Comment by kaldun-tech** [comment]
@hieblmi @ffranr @alexbosworth @starius @bhandras  Sorry for ping, reminder to review

**Comment by hieblmi** [comment]
@claude review this

**Comment by kaldun-tech** [comment]
The CI failure looks like a flake. Claude reports:

> It's a transient CI infrastructure issue with Go module fetching from GitLab-hosted dependencies (modernc.org/ccgo/v3). The error:
> fatal: shallow file has changed since we read it
> This happens when GitHub Actions' shallow clone conflicts with Go's attempt to fetch certain GitLab-hosted modules with GOPROXY=direct.
> Action: Re-run the failed job - this is flaky and often passes on retry
>
> ## Solution
> Add a step to the `mod-check` job in [`.github/workflows/main.yml`](https://github.com/lightninglabs/loop/blob/7529656f53972997c1abf684a23d5f56fc3c4038/.github/workflows/main.yml)
> to disable shallow clones for Go modules:

```yaml
mod-check:
name: go mod check
runs-on: ubuntu-latest
steps:
- name: git checkout
uses: actions/checkout@v4

- name: disable git shallow clone for go modules
run: git config --global --add safe.directory '*'

- name: setup go ${{ env.GO_VERSION }}
uses: actions/setup-go@v2
with:
go-version: '~${{ env.GO_VERSION }}'

- name: run check
run: make mod-check
```

> Alternatively, you can set an environment variable to work around the shallow clone issue:

```yaml
- name: run check
run: make mod-check
env:
GIT_CLONE_PROTECTION_ACTIVE: false
```

> The root cause is that the GitHub Actions checkout uses shallow clones by default for performance, which conflicts with Go's module fetching when dealing with certain Git repositories (like the GitLab-hosted `modernc.org/ccgo`). The fix ensures Git can properly handle the module dependencies during `go mod tidy`.

**Comment by kaldun-tech** [comment]
Addressing Claude bot's comments:
1. Adding String() -> "[redacted]": I made this design decision (see PR description) because
- The `DSN(hidePassword bool) `method handles the primary logging path (used in loopdb/postgres.go line 87)
- It can mask debugging issues when troubleshooting auth failures
- The underlying string is still accessible via cast, so it's not a security boundary
I can add the String() method if maintainers prefer defense-in-depth over debug-ability here
2. MarshalJSON protection: The config is CLI-parsed via go-flags and never serialized to JSON in normal operation. Adding MarshalJSON for a theoretical leak vector adds complexity IMO without clear benefit. I did not find any JSON marshalling of PostgresConfig anywhere. It's only used via go-flags CLI parsing.
3. No escape for @@ → literal @: Also acknowledged in PR desc as a known tradeoff. The workaround: if your password starts with @, store it in a file. Adding @@ escape syntax increases complexity and parsing edge cases. Can document this limitation more explicitly in the field description
4. File permission validation: This would be nice to have but adds platform-specific complexity with Windows vs Unix permissions. The primary security benefit of the Secret type feature is keeping secrets out of CLI args and config files. Not enforcing file permissions. Can be a follow-up enhancement PR
5. Confirm nolint removal - will verify with linter

File: loopdb/secret_test.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+		require.NoError(t, err)
+		require.Equal(t, Secret("secretpassword"), s)
+	})
+
+	t.Run("file with CRLF", func(t *testing.T) {
+		t.Parallel()
+
+		tmpDir := t.TempDir()
+		passFile := filepath.Join(tmpDir, "password.txt")
+		err := os.WriteFile(passFile, []byte("secretpassword\r\n"), 0600)
```
Comment (kaldun-tech) [inline loopdb/secret_test.go:68]: let me know if you would prefer to extract constants for reused values like "password.txt" or 0600

File: loopdb/secret_test.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+		require.NoError(t, err)
+		require.Equal(t, Secret(""), s)
+	})
+
+	t.Run("file with newline in middle", func(t *testing.T) {
+		t.Parallel()
+
+		tmpDir := t.TempDir()
+		passFile := filepath.Join(tmpDir, "password.txt")
+		err := os.WriteFile(passFile, []byte("pass\nword\n"), 0600)
```
Comment (kaldun-tech) [inline loopdb/secret_test.go:124]: this is an odd test case but I believe the correct behavior

**Comment by kaldun-tech** [comment]
@starius @bhandras Also pinging for review if you're available

**Comment by litbot-9000** [comment]
@hieblmi: review reminder
@kaldun-tech, remember to re-request review from reviewers when ready

**Comment by litbot-9000** [comment]
@hieblmi: review reminder
@kaldun-tech, remember to re-request review from reviewers when ready

**Comment by litbot-9000** [comment]
@hieblmi: review reminder
@kaldun-tech, remember to re-request review from reviewers when ready

**Comment by litbot-9000** [comment]
@hieblmi: review reminder
@kaldun-tech, remember to re-request review from reviewers when ready

**Comment by litbot-9000** [comment]
@hieblmi: review reminder
@kaldun-tech, remember to re-request review from reviewers when ready

**Comment by litbot-9000** [comment]
@hieblmi: review reminder
@kaldun-tech, remember to re-request review from reviewers when ready

**Comment by litbot-9000** [comment]
@hieblmi: review reminder
@kaldun-tech, remember to re-request review from reviewers when ready
